### PR TITLE
Use ncclAlltoAllv and ncclAlltoAll API when supported

### PR DESCRIPTION
**Pull request by dsjohns2**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134499

NCCL does not have an api for ncclAllToAll and ncclAllToAllv, so PyTorch does point to point send/recv. Expose this API if it is supported.

Differential Revision: [D61683836](https://our.internmc.facebook.com/intern/diff/D61683836/)

### CODE REVIEW

File: torch/csrc/cuda/nccl.cpp [hunk 1]
```diff
@@ -877,6 +877,18 @@ void all2all_single_unequal_split(
 
   auto type = to_nccl_data_type(_type);
   auto comm = to_nccl_comm(_comm);
+#ifdef NCCL_ALLTOALLV_SUPPORTED
```
Comment (dsjohns2) [inline torch/csrc/cuda/nccl.cpp:880]:
```suggestion
// NCCL_ALLTOALL_SUPPORTED is used so NCCL can differentiate send/recv
// operations issued as a part of the collective (e.g. alltoall) vs those inside traditional
// p2p operations.
#ifdef NCCL_ALLTOALLV_SUPPORTED
```

File: torch/csrc/cuda/nccl.cpp [hunk 2]
```diff
@@ -832,7 +832,7 @@ void all2all_single_equal_split(
   const auto* sendbuff = reinterpret_cast<const char*>(input.const_data_ptr());
   auto* recvbuff = reinterpret_cast<char*>(output.data_ptr());
   auto comm = to_nccl_comm(_comm);
-#if defined(USE_ROCM)
+#if defined(USE_ROCM) || defined(NCCL_ALLTOALL_SUPPORTED)
```
Comment (dsjohns2) [inline torch/csrc/cuda/nccl.cpp:835]:
```suggestion
// NCCL_ALLTOALL_SUPPORTED is used so NCCL can differentiate send/recv
// operations issued as a part of the collective (e.g. alltoall) vs those inside traditional
// p2p operations.
#if defined(USE_ROCM) || defined(NCCL_ALLTOALL_SUPPORTED)
```

File: torch/csrc/cuda/nccl.cpp [hunk 3]
```diff
@@ -924,6 +937,45 @@ void all2all(
   using namespace torch::cuda::nccl::detail;
   auto comm = to_nccl_comm(_comm);
 
+#ifdef NCCL_ALLTOALLV_SUPPORTED
```
Comment (dsjohns2) [inline torch/csrc/cuda/nccl.cpp:940]:
```suggestion
// NCCL_ALLTOALLV_SUPPORTED is used so NCCL can differentiate send/recv
// operations issued as a part of the collective (e.g. alltoallv) vs those inside traditional
// p2p operations.
#ifdef NCCL_ALLTOALLV_SUPPORTED
```

**Review by shuqiangzhang — APPROVED** [review]
There were discussions that PT should be able to allow ncclAllToAll to be implemented in different backends to improve perf and better visibility

**Review by eqy — APPROVED** [review]

**Pull request closed without merging**